temporal-sdk/src/main/java/io/temporal/common/converter/AbstractProtobufPayloadConverter.java

```
@@ -0,0 +1,79 @@
+import io.temporal.api.common.v1.Payload;
+import java.lang.reflect.Method;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractProtobufPayloadConverter {
+  protected static final Logger log =
+      LoggerFactory.getLogger(AbstractProtobufPayloadConverter.class);
+
+  private final boolean excludeProtobufMessageTypes;
+
+  protected AbstractProtobufPayloadConverter() {
+    this.excludeProtobufMessageTypes = false;
+  }
+
+  protected AbstractProtobufPayloadConverter(boolean excludeProtobufMessageTypes) {
+    this.excludeProtobufMessageTypes = excludeProtobufMessageTypes;
+  }
+
+  protected void addMessageType(Payload.Builder builder, Object value) {
+    if (this.excludeProtobufMessageTypes) {
+      return;
+    }
+    String messageTypeName = ((MessageOrBuilder) value).getDescriptorForType().getFullName();
+    builder.putMetadata(
+        EncodingKeys.METADATA_MESSAGE_TYPE_KEY, ByteString.copyFrom(messageTypeName, UTF_8));
+  }
+
+  protected void checkMessageType(Payload payload, Class valueClass) throws DataConverterException {
+    ByteString messageTypeBytes =
+        payload.getMetadataMap().get(EncodingKeys.METADATA_MESSAGE_TYPE_KEY);
+    if (messageTypeBytes != null) {
+      try {
+        String messageType = messageTypeBytes.toString(UTF_8);
+        Method getDescriptor = valueClass.getMethod("getDescriptor");
+        String valueMessageType =
+            ((Descriptors.Descriptor) getDescriptor.invoke(null)).getFullName();
+        if (!messageType.equals(valueMessageType) && log.isWarnEnabled()) {
+          log.warn(
+              "Encoded protobuf message type \""
+                  + messageType
+                  + "\" does not match valueClass `"
+                  + valueClass.getName()
+                  + "`, which has message type \""
+                  + valueMessageType
+                  + '"');
+        }
+      } catch (Exception e) {
+        throw new DataConverterException(e);
```

If it's just a log thing, let's never rethrow anything here. If an exception happens, let's swallow it and just print it in a log.

---

Makes sense, will do